homeassistant/components/history_stats/sensor.py

```diff
@@ -96,214 +96,72 @@ async def async_setup_platform(
-        # of the period
-        previous_state_matches = (
-            self._history_current_period
-            and self._history_current_period[0].state in self._entity_states
-        )
-        last_state_change_timestamp = start_timestamp
-        elapsed = 0.0
-        changes_to_match_state = 0
-
-        # Make calculations
-        for item in self._history_current_period:
-            current_state_matches = item.state in self._entity_states
-            state_change_timestamp = item.last_changed.timestamp()
-
-            if previous_state_matches:
-                elapsed += state_change_timestamp - last_state_change_timestamp
-            elif current_state_matches:
-                changes_to_match_state += 1
-
-            previous_state_matches = current_state_matches
-            last_state_change_timestamp = state_change_timestamp
-
-        # Count time elapsed between last history state and end of measure
-        if previous_state_matches:
-            measure_end = min(end_timestamp, now_timestamp)
-            elapsed += measure_end - last_state_change_timestamp
+class HistoryStatsSensor(HistoryStatsSensorBase):
+    """A HistoryStats sensor."""
 
-        # Save value in hours
-        hours_matched = elapsed / 3600
-        return hours_matched, changes_to_match_state
-
-    def _async_set_native_value(
-        self, hours_matched: float | None, changes_to_match_state: int | None
+    def __init__(
+        self,
+        coordinator: HistoryStatsUpdateCoordinator,
+        sensor_type: str,
+        name: str,
     ) -> None:
-        """Set attrs from value and count."""
-        if hours_matched is None:
+        """Initialize the HistoryStats sensor."""
+        super().__init__(coordinator, name)
+        self._attr_native_unit_of_measurement = UNITS[sensor_type]
+        self._type = sensor_type
+
+    @callback
+    def _process_update(self) -> None:
+        """Process an update from the coordinator."""
+        state = self.coordinator.data
+        if state is None or state.hours_matched is None:
             self._attr_native_value = None
             self._attr_extra_state_attributes = {}
             return
 
         if self._type == CONF_TYPE_TIME:
-            self._attr_native_value = round(hours_matched, 2)
+            self._attr_native_value = round(state.hours_matched, 2)
         elif self._type == CONF_TYPE_RATIO:
-            self._attr_native_value = HistoryStatsHelper.pretty_ratio(
-                hours_matched, self._period
-            )
+            self._attr_native_value = pretty_ratio(state.hours_matched, state.period)
         elif self._type == CONF_TYPE_COUNT:
-            self._attr_native_value = changes_to_match_state
+            self._attr_native_value = state.changes_to_match_state
         self._attr_extra_state_attributes = {
-            ATTR_VALUE: HistoryStatsHelper.pretty_duration(hours_matched)
+            ATTR_VALUE: pretty_duration(state.hours_matched)
```

Comment:
I thought that we would drop the pretty duration ? or is that for once the frontend supports DURATION device class?

Reply:
I'll drop it as soon as the frontend supports the duration device class

Reply:
https://github.com/home-assistant/core/pull/70720